File: client/src/store/todos/usecase/create-todo/create-todo-interactor.ts
import { CreateTodoResponseModel } from './create-todo-responsemodel';
import { TodoGateway } from '@/adapter/gateway/todo-gateway';
import { CreateTodoRequestModel } from './create-todo-requestmodel';

export class CreateTodoInteractor {
  private todoGateway: TodoGateway;
  constructor(todoGateway: TodoGateway) {
    this.todoGateway = todoGateway;
  }
  async createTodo(
    requestModel: CreateTodoRequestModel
  ): Promise<CreateTodoResponseModel> {
    const { todo, currentTodos } = requestModel;

    const newTodo = await this.todoGateway.createTodo(todo);
    const newTodos = { ...currentTodos, newTodo };

    return { todos: newTodos };
  }
}

File: server/usecase/createtodo/createtodo_interactor.go
package createtodo

import "github.com/kenfdev/cleancoding/server/adapter/gateway"

type createTodoInteractor struct {
	todoGateway gateway.TodoGateway
}

func NewCreateTodoInteractor(tdGateway gateway.TodoGateway) CreateTodoInputPort {
	return &createTodoInteractor{
		todoGateway: tdGateway,
	}
}

func (it *createTodoInteractor) CreateTodo(requestModel *CreateTodoRequestModel, outputPort CreateTodoOutputPort) error {
	todo, err := it.todoGateway.CreateTodo(requestModel.Todo)
	if err != nil {
		return err
	}

	res := &CreateTodoResponseModel{
		Todo: todo,
	}

	outputPort.Present(res)

	return nil
}
